piq/ssim.py

```diff
@@ -433,48 +362,44 @@ def _ssim_per_channel(x: torch.Tensor, y: torch.Tensor, kernel: torch.Tensor, da
         Full Value of Structural Similarity (SSIM) index.
     """
 
-    if x.size(-1) < kernel.size(-1) or x.size(-2) < kernel.size(-1):
+    if x.size(-1) < kernel.size(-1) or x.size(-2) < kernel.size(-2):
         raise ValueError(f'Kernel size can\'t be greater than actual input size. Input size: {x.size()}. '
                          f'Kernel size: {kernel.size()}')
 
     c1 = (k1 * data_range) ** 2
     c2 = (k2 * data_range) ** 2
-
-    kernel = kernel.to(x.device, dtype=x.dtype)
-
-    mu1 = _gaussian_filter(x, kernel)
-    mu2 = _gaussian_filter(y, kernel)
+    n_channels = x.size(1)
+    mu1 = f.conv2d(x, weight=kernel, stride=1, padding=0, groups=n_channels)
+    mu2 = f.conv2d(y, weight=kernel, stride=1, padding=0, groups=n_channels)
 
     mu1_sq = mu1.pow(2)
     mu2_sq = mu2.pow(2)
     mu1_mu2 = mu1 * mu2
 
     compensation = 1.0
-    sigma1_sq = compensation * (_gaussian_filter(x * x, kernel) - mu1_sq)
-    sigma2_sq = compensation * (_gaussian_filter(y * y, kernel) - mu2_sq)
-    sigma12 = compensation * (_gaussian_filter(x * y, kernel) - mu1_mu2)
+    sigma1_sq = compensation * (f.conv2d(x * x, weight=kernel, stride=1, padding=0, groups=n_channels) - mu1_sq)
+    sigma2_sq = compensation * (f.conv2d(y * y, weight=kernel, stride=1, padding=0, groups=n_channels) - mu2_sq)
+    sigma12 = compensation * (f.conv2d(x * y, weight=kernel, stride=1, padding=0, groups=n_channels) - mu1_mu2)
 
     # Set alpha = beta = gamma = 1.
     cs_map = (2 * sigma12 + c2) / (sigma1_sq + sigma2_sq + c2)
-    ssim_map = ((2 * mu1_mu2 + c1) / (mu1_sq + mu2_sq + c1)) * cs_map
+    ssim_map = (2 * mu1_mu2 + c1) / (mu1_sq + mu2_sq + c1) * cs_map
```

Comment:
I understand that removing these parentecies are redundant since they do not change the order of operations. However, I would still leave them to make this order more explicit.

Reply:
I suggest to keep it as is. Pytorch 1.6.0 will support complex values, which will lead to possible refactoring of the logic for complex valued methods.